common/Obelisk/OAuth/Request.hs

```diff
@@ -0,0 +1,124 @@
+import           Data.Text             (Text)
+import qualified Data.Text             as T
+
+import           Obelisk.OAuth.Route   (OAuthClientId (..), OAuthRoute (..),
+                                        redirectUriParamsEncoder)
+import           Obelisk.OAuth.State   (OAuthState, oAuthStateAsText)
+import           Obelisk.Route
+
+import           Obelisk.OAuth.Config  (AuthorizationResponseType (..),
+                                        OAuthConfig (..))
+
+
+-- | Render the authorization request.
+--
+--   This should be all you need from this module in most cases.
+authorizationRequestHref
+  :: Encoder Identity Identity (R (Sum r a)) PageName -- ^ Backend route encoder
+  -> OAuthConfig secret r -- ^ Configuration for building up request.
+  -> OAuthState -- ^ The state for building the request URI.
+  -> Text
+     -- ^ Authorization grant request endpoint with query string
+authorizationRequestHref enc cfg state =
+  _oAuthConfig_providerUri cfg <> "?" <> authorizationRequestParams enc cfg state
+
+
+-- | Turn an 'AuthorizationRequest' into query string parameters separated by
+--   @&@. Key names are defined in
+--
+-- <https://tools.ietf.org/html/rfc6749#section-4.1.1 4.1.1> of the
+-- specification.  This does not insert a leading @?@.
+authorizationRequestParams
+  :: Encoder Identity Identity (R (Sum r a)) PageName
+     -- ^ Encoder for `_oAuthConfig_redirectUri`
+  -> OAuthConfig secret r
+     -- ^ Configruation for building parameters
+  -> OAuthState
+  -> Text
+authorizationRequestParams enc cfg state = encode (queryParametersTextEncoder @Identity @Identity) $
+  Map.toList $ fmap Just $ mconcat
+    [ Map.singleton "response_type" $ case _oAuthConfig_responseType cfg of
+        AuthorizationResponseType_Code  -> "code"
+        AuthorizationResponseType_Token -> "token"
+    , Map.singleton "client_id" (unOAuthClientId . _oAuthConfig_clientId $ cfg)
+    , case _oAuthConfig_redirectUri cfg of
+        Nothing -> Map.empty
+        Just (b, r) -> Map.singleton "redirect_uri" $
+          renderRedirectUri enc b r
+    , case _oAuthConfig_scope cfg of
+        [] -> Map.empty
+        xs -> Map.singleton "scope" $ T.unwords xs
+    , Map.singleton "state" (oAuthStateAsText state)
+    ]
+
+
+-- | Given the base URL  of the client application, a checked backend route
+--   encoder (see 'checkEncoder'), render the redirect URI.
+renderRedirectUri
+  :: Encoder Identity Identity (R (Sum r a)) PageName -- ^ Checked backend route encoder
+  -> Text -- ^ Application base url
+  -> r (R OAuthRoute) -- ^ OAuth parent route
+  -> Text -- ^ Rendered redirect url
+renderRedirectUri enc base = (base <>) . renderRedirectUriRoute enc
+
+
+-- | Given a checked backend route encoder (see 'checkEncoder'), and the
+--   app-specific parent route under which the OAuth routes are nested, construct
+--   the redirect URI's route (i.e., the path and query string parts).
+renderRedirectUriRoute
```

Comment:
This function's interface and documentation are considerably longer than its implementation.  Is it really helping to break it out as a separate function, or is everyone using it just going to need to read the implementation anyhow?